source/extensions/filters/network/dubbo_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,325 @@
+
+  Upstream::ThreadLocalCluster* cluster = cluster_manager_.get(route_entry_->clusterName());
+  if (!cluster) {
+    ENVOY_STREAM_LOG(debug, "dubbo router: unknown cluster '{}'", *callbacks_,
+                     route_entry_->clusterName());
+    callbacks_->sendLocalReply(
+        AppException(ResponseStatus::ServerError, fmt::format("dubbo router: unknown cluster '{}'",
+                                                              route_entry_->clusterName())),
+        true);
+    return Network::FilterStatus::StopIteration;
+  }
+
+  cluster_ = cluster->info();
+  ENVOY_STREAM_LOG(debug, "dubbo router: cluster '{}' match for interface '{}'", *callbacks_,
+                   route_entry_->clusterName(), metadata->service_name());
+
+  if (cluster_->maintenanceMode()) {
+    callbacks_->sendLocalReply(
+        AppException(ResponseStatus::ServerError,
+                     fmt::format("dubbo router: maintenance mode for cluster '{}'",
+                                 route_entry_->clusterName())),
+        true);
+    return Network::FilterStatus::StopIteration;
+  }
+
+  Tcp::ConnectionPool::Instance* conn_pool = cluster_manager_.tcpConnPoolForCluster(
+      route_entry_->clusterName(), Upstream::ResourcePriority::Default, this, nullptr);
+  if (!conn_pool) {
+    callbacks_->sendLocalReply(
+        AppException(
+            ResponseStatus::ServerError,
+            fmt::format("dubbo router: no healthy upstream for '{}'", route_entry_->clusterName())),
+        true);
+    return Network::FilterStatus::StopIteration;
+  }
+
+  ENVOY_STREAM_LOG(debug, "dubbo router: decoding request", *callbacks_);
+
+  upstream_request_ = std::make_unique<UpstreamRequest>(*this, *conn_pool, metadata,
+                                                        callbacks_->downstreamSerializationType(),
+                                                        callbacks_->downstreamProtocolType());
+  return upstream_request_->start();
+}
+
+void Router::onUpstreamData(Buffer::Instance& data, bool end_stream) {
+  ASSERT(!upstream_request_->response_complete_);
+
+  ENVOY_STREAM_LOG(trace, "dubbo router: reading response: {} bytes", *callbacks_, data.length());
+
+  // Handle normal response.
+  if (!upstream_request_->response_started_) {
+    callbacks_->startUpstreamResponse(*upstream_request_->deserializer_.get(),
+                                      *upstream_request_->protocol_.get());
+    upstream_request_->response_started_ = true;
+  }
+
+  DubboFilters::UpstreamResponseStatus status = callbacks_->upstreamData(data);
+  if (status == DubboFilters::UpstreamResponseStatus::Complete) {
+    ENVOY_STREAM_LOG(debug, "dubbo router: response complete", *callbacks_);
+    upstream_request_->onResponseComplete();
+    cleanup();
+    return;
+  } else if (status == DubboFilters::UpstreamResponseStatus::Reset) {
+    ENVOY_STREAM_LOG(debug, "dubbo router: upstream reset", *callbacks_);
+    upstream_request_->resetStream();
```

Comment:
Why isn't onResponseComplete called here?

Reply:
Reset merely resets the upstream connection and does not indicate that the request has been completed.